frontend/src/app/max-validator.directive.ts

```diff
@@ -0,0 +1,50 @@
+/* tslint:disable:directive-selector */
+import { Directive, forwardRef, HostBinding, Input, OnChanges, SimpleChanges } from '@angular/core';
+import { AbstractControl, NG_VALIDATORS, ValidationErrors, Validator, ValidatorFn, Validators } from '@angular/forms';
+
+/**
+ * Provider which adds MaxValidatorDirective to NG_VALIDATORS.
+ */
+export const MAX_VALIDATOR: any = {
+  provide: NG_VALIDATORS,
+  useExisting: forwardRef(() => MaxValidatorDirective),
+  multi: true
+};
+
+/**
+ * A directive which installs the MaxValidator for any `formControlName`,
+ * `formControl`, or control with `ngModel` that also has a `max` attribute.
+ */
+@Directive({
+  selector: '[max][type=number][formControlName],[max][type=number][formControl],[max][type=number][ngModel]',
+  providers: [MAX_VALIDATOR]
+})
+export class MaxValidatorDirective implements Validator, OnChanges {
+  private _validator: ValidatorFn;
+  private _onChange: () => void;
+
+  @Input() max: number;
+
+  @HostBinding('attr.max') get maxAttr() {
+    return this.max !== null && this.max !== undefined ? `${this.max}` : null;
+  }
+
+  ngOnChanges(changes: SimpleChanges): void {
+    if ('max' in changes) {
+      this._createValidator();
+      if (this._onChange) {
+        this._onChange();
+      }
+    }
+  }
+
+  validate(c: AbstractControl): ValidationErrors | null {
+    return this.max == null ? null : this._validator(c);
+  }
+
+  registerOnValidatorChange(fn: () => void): void { this._onChange = fn; }
+
+  private _createValidator(): void {
+    this._validator = Validators.max(this.max);
```

Comment:
I see that the original directive were using `parseInt` because `max` was a string.
Don't you have the same problem? I don't know if browsers still do the weird parseInt with octal radix thing...

Reply:
I saw that, too. I find it strange: if you use `[max]="someValue"`, you pass a number, not a string. But yes, indeed, if you use `max="0"`, you pass a string. So whetever the type you use, you lie to yourself in one of the cases, and it only works because JavaScript transparently converts the number to a string or vice-versa. So I chose what I find the most logical: the type number. I don't think it makes a difference, except when the original directive uses 

    host: {'[attr.minlength]': 'minlength ? minlength : null'}

which is probably a bug, because if the minlength actual runtime type is a number, and its value is 0, then the attribute is not added anymore. Not really a problem for minlength, since the value 0 is meaningless. But for min and max, it's more important. Demo here: http://plnkr.co/edit/YrH5D2qikr5umOSU9xkI?p=preview

If you inspect the second input, it doesn't have a minlength. If you inspect the third one, it has one.